srtcore/buffer_rcv.cpp

```diff
@@ -159,53 +167,257 @@ int CRcvBuffer::insert(CUnit* unit)
+    // this comparison a simple subtraction. Note that offset will
+    // have to be updated on every shift of m_iStartPos.
+    else if (cmpPos(newpktpos, m_iDropPos) < 0)
+    {
+        // Case [C]: the newly inserted packet precedes the
+        // previous earliest delivery position after drop,
+        // that is, there is now a "better" after-drop delivery
+        // candidate.
+
+        // New position updated a valid packet on an earlier
+        // position than the drop position was before, although still
+        // following a gap.
+        //
+        // We know it because if the position has filled a gap following
+        // a valid packet, this preceding valid packet would be pointed
+        // by m_iDropPos, or it would point to some earlier packet in a
+        // contiguous series of valid packets following a gap, hence
+        // the above condition wouldn't be satisfied.
+        m_iDropPos = newpktpos;
+
+        // If there's an inserted packet BEFORE drop-pos (which makes it
+        // a new drop-pos), while the very first packet is absent (the
+        // below condition), it means we have a new earliest-available
+        // packet. Otherwise we would have only a newly updated drop
+        // position, but still following some earlier contiguous range
+        // of valid packets - so it's earlier than previous drop, but
+        // not earlier than the earliest packet.
+        if (m_iStartPos == m_iEndPos)
+        {
+            earlier_time = getPktTsbPdTime(unit->m_Packet.getMsgTimeStamp());
+        }
+    }
+    // OTHERWISE: case [D] in which nothing is to be updated.
+
     // If packet "in order" flag is zero, it can be read out of order.
     // With TSBPD enabled packets are always assumed in order (the flag is ignored).
     if (!m_tsbpd.isEnabled() && m_bMessageAPI && !unit->m_Packet.getMsgOrderFlag())
     {
-        ++m_numOutOfOrderPackets;
-        onInsertNotInOrderPacket(pos);
+        ++m_numNonOrderPackets;
+        onInsertNonOrderPacket(newpktpos);
     }
 
     updateNonreadPos();
+
+    CPacket* avail_packet = NULL;
+
+    if (m_entries[m_iStartPos].pUnit && m_entries[m_iStartPos].status == EntryState_Avail)
+    {
+        avail_packet = &packetAt(m_iStartPos);
+        avail_range = offPos(m_iStartPos, m_iEndPos);
+    }
+    else if (!m_tsbpd.isEnabled() && m_iFirstNonOrderMsgPos != -1)
+    {
+        // In case when TSBPD is off, we take into account the message mode
+        // where messages may potentially span for multiple packets, therefore
+        // the only "next deliverable" is the first complete message that satisfies
+        // the order requirement.
+        avail_packet = &packetAt(m_iFirstNonOrderMsgPos);
+        avail_range = 1;
+    }
+    else if (m_iDropPos != m_iEndPos)
+    {
+        avail_packet = &packetAt(m_iDropPos);
+        avail_range = 1;
+    }
+    else
+    {
+        avail_packet = NULL;
+        avail_range = 0;
+    }
```

Comment:
Consider moving to a separate function.
Maybe even the whole logic around forming the `InsertPos` may be consolidated in a single function.

Reply:
Good idea.